### PR TITLE
Fail Gracefully

### LINKED ISSUE #14: pvl should be able to fail gracefully when a label doesn't parse correctly
`pvl` should be able to fail gracefully when a label doesn't parse correctly.  Here is a sample product that fails:

http://pds-imaging.jpl.nasa.gov/data/mer1-m-pancam-5-anaglyph-ops-v1.0/mer1po_0xxx/data/sol0072/rdr/1p134581832rsl0932p2415r2m1.img

For elements that don't parse, it should just carry on and return what it can of the label.  I am not quite sure what to do with the failed element though.

The example above fails because it lacks a value altogether (note the ^M are carriage returns displayed by VIM, they are fine) ...

```
GROUP                             = DERIVED_IMAGE_PARMS^M
  RADIANCE_OFFSET                 = 0.0 <WATT*M**-2*SR**-1*NM**-1>^M
  RADIANCE_SCALING_FACTOR         = 1.656e-05 <WATT*M**-2*SR**-1*NM**-1>^M
  RADIOMETRIC_CORRECTION_TYPE     = MIPLRAD2^M
  DERIVED_IMAGE_TYPE              = ^M
END_GROUP                         = DERIVED_IMAGE_PARMS^M
```

### PR DESCRIPTION
**Pull request by pbvarga1**
Fixes #14 

Make it so when the label is broken (i.e. there is a key without a value), pvl will still read the entire label and not fail.

### CODE REVIEW

**Comment by pbvarga1** [comment]
This is definitely still a work in progress but wanted to start the discussion here on how I should proceed. As of b081e29, the empty portion will just be an empty string ``''``. I know there was discussion of somehow storing the error somewhere but I haven't looked at that portion yet.

File: tests/test_decoder.py [hunk 1]
```diff
@@ -734,3 +734,111 @@ def test_parse_error():
 
     with pytest.raises(pvl.decoder.ParseError):
         pvl.load(io.BytesIO(b'foo'))
+
+
+@pytest.mark.parametrize(
```
Comment (pbvarga1) [inline tests/test_decoder.py]: I've come up with 16 broken test labels. I have to spend more time figuring out edge cases and I will probably move these to their own test files later.

File: tests/test_decoder.py [hunk 2]
```diff
@@ -714,17 +714,17 @@ def test_end_comment():
 
 
 def test_parse_error():
-    with pytest.raises(pvl.decoder.ParseError):
```
Comment (pbvarga1) [inline tests/test_decoder.py]: ``'foo='`` is now a valid label

Comment (godber) [inline tests/test_decoder.py]: But is `foo=` actually a valid label according to the PVL specification?

http://pirlwww.lpl.arizona.edu/software/PPVL/CCSDS-641.0-B-2.pdf

I don't think so.  So I think the `pvl` module needs to have two modes, a strict mode where it actually raises an error for truly bad labels and another mode where it is permissive and does what you're doing.  `mode='strict'` and `mode='permissive'`.

The default should be `strict` and the error message raised on error should be unique and prominent enough in the docs that when users google it, it shows up in our docs and then can then switch to `mode='permissive'`.

Comment (pbvarga1) [inline tests/test_decoder.py]: Ok I agree.  I'll implement the ``mode`` in a couple commits. But I think instead of ``mode='strict'`` it should be a boolean like ``strict=True``. I think that will be easier to remember and still as clear as ``mode``

**Comment by pbvarga1** [comment]
Commit d270ef4 creates ``EmptyValue`` for the parameters without values
Example:
```python
>>> import pvl
>>> label = """
...         foo = bar
...         life =
...         monty = python
...         """
>>> module = pvl.loads(label)
>>> module
PVLModule([
  (u'foo', u'bar')
  (u'life', EmptyValue(Line 3 does not have a value. Treat as an empty string))
  (u'monty', u'python')
])
>>> module['life']
EmptyValue(Line 3 does not have a value. Treat as an empty string)
>>> module['life'].lineno
3
>>> 'ex' + module['life']
'ex'
>>> int(module['life'])
0
>>> float(module['life'])
0.0
```

**Comment by pbvarga1** [comment]
Commit 3ca26d5 creates ``strict=True`` kwarg
Example:
```python
>>> import pvl
>>> label = """
...         foo = bar
...         life =
...         monty = python
...         """
>>> module = pvl.loads(label)
ParseError: Broken Parameter-Value. Set strict to False to skip error: line 3
>>> module = pvl.loads(label, strict=False)
>>> module
PVLModule([
  (u'foo', u'bar')
  (u'life', EmptyValue(Line 3 does not have a value. Treat as an empty string))
  (u'monty', u'python')
])
>>> module['life']
EmptyValue(Line 3 does not have a value. Treat as an empty string)
>>> module['life'].errors
[3]
```

**Comment by pbvarga1** [comment]
As is, I think this is a stable complete fix to #14. I've been trying to think of more test cases but everything I come up with either is just another iteration of a previous test or is too broken and ambiguous of a label that the user just needs to actually fix the label instead. All the problematic labels I have run into have been due to parameters not having a value and this pull request fixes that and make it un-ambiguous that there was a problematic parameter-value pair.

File: pvl/decoder.py [hunk 3]
```diff
@@ -16,14 +16,38 @@ class ParseError(ValueError):
...
+            errmsg = '%s: line %d' % (msg, lineno)
         super(ParseError, self).__init__(errmsg)
         self.msg = msg
         self.pos = pos
         self.lineno = lineno
         self.colno = colno
 
 
+class EmptyValue(str):
+
```
Comment (godber) [inline pvl/decoder.py]: This class might benefit from a docstring describing why it exists and how to use it (or how it's used).  For instance, I see these `EV(1)`, `EV(2)`, and `EV(3)`.   I don't get what the numbers are.

File: pvl/decoder.py [hunk 4]
```diff
@@ -109,6 +137,14 @@ def raise_unexpected(self, stream, token=None):
     def raise_unexpected_eof(self, stream):
         self.raise_error('Unexpected EOF', stream)
 
+    def broken_parameter_value(self, lineno):
```
Comment (godber) [inline pvl/decoder.py]: Is "parameter_value" the best name for the thing that is broken?  Does the PVL spec use the words "parameter value"?  It would be best to stick to the language in the spec if possible

File: pvl/decoder.py [hunk 5]
```diff
@@ -109,6 +137,14 @@ def raise_unexpected(self, stream, token=None):
...
+        if self.strict:
+            msg = "Broken Parameter-Value. Set strict to False to skip error"
```
Comment (godber) [inline pvl/decoder.py]: After (possibly) addressing the naming issue above, consider rewording this message to be more explicit.  Your message tells the user what to do to make this work, but could be reworded so that they know exactly what to do.  Try:

"Broken Parameter-Value.  Using 'strict=False' when calling 'pvl.load' may help you parse the label, it could also inadvertently mask other errors."

The "inadvertently masking errors" bit is the warning to the user that using the strict option might not be a great idea.  Which it may not be.  It's our disclaimer.

File: tests/test_decoder.py [hunk 6]
```diff
@@ -734,3 +736,132 @@ def test_parse_error():
 
     with pytest.raises(pvl.decoder.ParseError):
         pvl.load(io.BytesIO(b'foo'))
+
```
Comment (godber) [inline tests/test_decoder.py]: Can you reproduce the following issue?

I've added the following to `test_decoder.py`

```
BAD_PDS_LABELS = glob.glob(os.path.join(BROKEN_DIR, "*.lbl"))

def test_load_all_bad_sample_labels():
    for filename in BAD_PDS_LABELS:
        label = pvl.load(filename, strict=False)
        assert isinstance(label, Label)
```

And it fails with the error:

```

    def test_load_all_bad_sample_labels():
        for filename in BAD_PDS_LABELS:
>           label = pvl.load(filename, strict=False)

tests/test_decoder.py:704:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
pvl/__init__.py:89: in load
    return cls(**kwargs).decode(fp)
pvl/decoder.py:187: in decode
    module = PVLModule(self.parse_block(stream, self.has_end))
pvl/decoder.py:206: in parse_block
    statement = self.parse_statement(stream)
pvl/decoder.py:264: in parse_statement
    return self.broken_parameter_value(stream.lineno - 1)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <pvl.decoder.PVLDecoder object at 0x10b73aa90>, lineno = 2

    def broken_parameter_value(self, lineno):
        if self.strict:
            msg = "Broken Parameter-Value. Set strict to False to skip error"
>           raise ParseError(msg, None, lineno, None)
E           pvl.decoder.ParseError: Broken Parameter-Value. Set strict to False to skip error: line 2

pvl/decoder.py:143: ParseError
```

This is basically just comparing:

```
>>> import pvl
>>> a = pvl.load('tests/data/pds3/tiny4.lbl')
>>> b1 = pvl.load('tests/data/pds3/broken/broken1.lbl', strict=False)
```

I have that usage right yeah:  `strict=False`?

Comment (pbvarga1) [inline tests/test_decoder.py]: Exposed in 1f8f8aa and fixed in 700da87. It was a really simple and small bug. I just forgot to add the argument to the decoder when loading a filepath in [line 89](https://github.com/planetarypy/pvl/pull/29/commits/700da877b73d3819118fb6cbc0307b753226b074#diff-0a5de9c15bd399f5f184838723f4b310R89). Tests pass.

File: pvl/decoder.py (hunk 4 above)
Comment (pbvarga1) [inline pvl/decoder.py]: A better name would be ``broken_assignment`` but Parameter (more specifically Parameter Name) and Value to come from  the PVL spec.

File: pvl/decoder.py [hunk 7]
```diff
@@ -77,6 +137,13 @@ class PVLDecoder(object):
     decimal_chars = char_set('0123456789')
     hex_chars = char_set('0123456789ABCDEFabcdef')
 
+    def __init__(self):
```
Comment (pbvarga1) [inline pvl/decoder.py]: @godber I got rid of the keyword argument because I think that will preserve backwards compatibility for anyone that is making a subclass in order to decode

**Comment by godber** [comment]
You want to merge this or you want me to merge it?

**Comment by pbvarga1** [comment]
Should I change the version to 0.3.0 or do that in a separate pull request when I release the new version?

**Comment by pbvarga1** [comment]
Ill merge and update the version in a separate pull request
